src/tools/build-manifest/src/main.rs

```diff
@@ -375,6 +378,25 @@ impl Builder {
         }
     }
 
+    /// If a tool does not pass its tests, don't ship it.
+    /// Right now, we do this only for Miri.
+    fn check_toolstate(&mut self) {
+        let toolstates: Option<HashMap<String, String>> = try {
+            let toolstates = File::open(self.input.join("toolstates-linux.json")).ok()?;
+            serde_json::from_reader(&toolstates).ok()?
+        };
+        let toolstates = toolstates.unwrap_or_else(|| {
+            println!("WARNING: `toolstates-linux.json` missing; assuming all tools failed");
+            HashMap::default() // Use empty map if anything went wrong.
```

Comment:
Why did you choose this behavior instead of not doing anything if the file is missing/malformed?

Reply:
If we don't do anything we'll probably never notice that reading the file failed. This seemed like the safer default.

Reply:
That's fair.